### PR TITLE
[MAINTENANCE] Stack `RuleBasedProfiler` progress bars for better user experience

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Loom: https://www.loom.com/share/deb564f294f247f7a4a1b74d2695cd96
- Stacked progress bars for better readability and user experience

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Comment by ghost** [comment]
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=5400&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=5400&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)
</details>

[Review these changes using an interactive CodeSee Map](https://app.codesee.io/r/reviews?pr=5400&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
CodeSee Map Legend

File: scripts/check_type_hint_coverage.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 from typing import Dict, List, Optional
 
 TYPE_HINT_ERROR_THRESHOLD: int = (
-    2529  # This number is to be reduced as we annotate more functions!
```
Comment (Shinnnyshinshin) [inline scripts/check_type_hint_coverage.py:7]: ❤️

**Review by Shinnnyshinshin — APPROVED** [review]
Thank you for the Loom @cdkini, this is fantastic work :)

File: great_expectations/rule_based_profiler/rule/rule.py [hunk 2]
```diff
@@ -134,8 +137,16 @@ def run(
 
         rule_state.reset_parameter_containers()
 
+        pbar_method: Callable = determine_progress_bar_method_by_environment()
+
         domain: Domain
-        for domain in domains:
+        for domain in pbar_method(
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/rule/rule.py:143]: ❤️

File: great_expectations/rule_based_profiler/rule_based_profiler.py [hunk 3]
```diff
@@ -282,20 +284,21 @@ def run(
...
-        progress_bar.refresh()
         sys.stdout.write("\n")
         sys.stdout.flush()
 
+        pbar_method: Callable = determine_progress_bar_method_by_environment()
+
         rule_state: RuleState
         rule: Rule
-        for rule in effective_rules:
+        for rule in pbar_method(
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/rule_based_profiler.py:294]: ❤️

**Review by alexsherstinsky — APPROVED** [review]
LGTM!

**Pull request merged**